File: src/foam/crypto/hash/Hashable.js
/**
 * @license
 * Copyright 2018 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.INTERFACE({
  package: 'foam.crypto.hash',
  name: 'Hashable',

  documentation: 'Hashable interface',

  methods: [
    {
      name: 'hash',
      returns: 'ByteArray',
      javaThrows: [
        'java.security.NoSuchAlgorithmException',
      ],
      args: [
        {
          name: 'algorithm',
          type: 'String',
          documentation: 'Hashing algorithm to use'
        },
        {
          name: 'hash',
          type: 'ByteArray',
          documentation: 'Previous hash (used for chaining)'
        }
      ]
    }
  ]
});

File: src/foam/crypto/hash/Hashable.java
/**
 * @license
 * Copyright 2018 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package foam.crypto.hash;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public interface Hashable {

  /**
   * Hash method that uses default algorithm of SHA-256
   *
   * @return output of the hash function
   * @throws NoSuchAlgorithmException
   */
  byte[] hash()
    throws NoSuchAlgorithmException;

  /**
   * Hash method that uses a user provided hashing algorithm
   *
   * @param algorithm user provided hashing algorithm
   * @return output of the hash function
   * @throws NoSuchAlgorithmException
   */
  byte[] hash(String algorithm)
    throws NoSuchAlgorithmException;

  /**
   * Hash metohd that uses a user provided MessageDigest object
   * 
   * @param md MessageDigest object used for hashing
   * @return output of the hash function
   */
  byte[] hash(MessageDigest md);

}
